### PR TITLE
lang: funcs: Validates mac address strings

### PR DESCRIPTION
**Pull request by S-ign**
## Tips:

* please read the style guide before submitting your patch:
[docs/style-guide.md](../docs/style-guide.md)

* commit message titles must be in the form:

```topic: Capitalized message with no trailing period```

or:

```topic, topic2: Capitalized message with no trailing period```

* golang code must be formatted according to the standard, please run:

```
make gofmt		# formats the entire project correctly
```

or format a single golang file correctly:

```
gofmt -w yourcode.go
```

* please rebase your patch against current git master:

```
git checkout master
git pull origin master
git checkout your-feature
git rebase master
git push your-remote your-feature
hub pull-request	# or submit with the github web ui
```

* after a patch review, please ping @purpleidea so we know to re-review:

```
# make changes based on reviews...
git add -p		# add new changes
git commit --amend	# combine with existing commit
git push your-remote your-feature -f
# now ping @purpleidea in the github PR since it doesn't notify us automatically
```

### CODE REVIEW

**Comment by S-ign** [comment]
tests passed

**Review by purpleidea — CHANGES_REQUESTED** [review]
Hey @S-ign ! Thanks for the patch.

Structurally everything looks good! I left some comments. Make sure to review each one. If there's something you don't understand or agree with, leave a comment instead of updating things in your new patch.

For small patches, I generally recommend fixing things and force-pushing a new version.
For enormous patch sets, it's usually better to add a new commit with the fixes, and then rebase squash right at the end. (We have the privilege of hiding all of our mistakes before we commit them to git master forever!)

As for the commit message, try re-wording it and adding a small sentence or two describing anything unusual or special about this patch. Have a go.

One last tip: Usually the flow most people are used to looks like: 1) hack, hack, hack; 2) commit 3) push

I recommend adding a step 2.5 which involves reading your entire PR as if you were doing a review. This is helpful to both get better at reading code, and to make sure you have everything as you want it.

Thanks again!

File: lang/funcs/core/net/is_macaddress_func.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+)
+
+func init() {
+	simple.ModuleRegister(ModuleName, "ismacaddress", &types.FuncValue{
+		T: types.NewType("func(a str) bool"),
+		V: IsMacAddress,
+	})
+}
+
+// 00:21:cc:be:e0:6b
```
[the remark this answers is not in the document]
Comment (S-ign) [reply lang/funcs/core/net/is_macaddress_func.go:34]: yeah, that was a mistake, I put it there so I can just visually see a mac address while I was thinking about how to make the function.

File: lang/funcs/core/net/is_macaddress_func.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+func validate(mac, delimiter string) bool {
+	// valid delimiters
+	delims := map[string]bool{
+		":": true,
+		"-": true,
+		".": true,
+	}
+	if !delims[delimiter] {
```
[the remark this answers is not in the document]
Comment (S-ign) [reply lang/funcs/core/net/is_macaddress_func.go:54]: I tried that out, if delimiter is not in the map it will be false and since the logic statement its asking for the reverse of that, it will be true, once inside the statement it'll then return false as it should because "X" isn't one of the accepted delimiters, and the mac address string will be invalid as well. I'm not sure on a safe way to do this though.

File: lang/funcs/core/net/is_macaddress_func_test.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		{"invalid mac invalid delimiter position", "0a::1b2:c3d:4f:56", false, nil},
+		{"invalid mac invalid delimiter position", "0a--1b2-c3d-4f-56", false, nil},
+	}
+
+	for _, test := range macAddressTests {
+		t.Run(test.name, func(t *testing.T) {
+			output, err := IsMacAddress([]types.Value{
+				&types.StrValue{V: test.mac},
+			})
+			expectedStr := &types.BoolValue{V: test.expected}
```
[the remark this answers is not in the document]
Comment (S-ign) [reply lang/funcs/core/net/is_macaddress_func_test.go:55]: ok, so it has to be string, thought I could return other types too.

File: lang/funcs/core/net/is_macaddress_func_test.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+			if test.err != nil && err.Error() != test.err.Error() {
+				t.Errorf("mac %s, expected error: %q, got %q", test.mac, test.err, err)
```
[the remark this answers is not in the document]
Comment (S-ign) [reply lang/funcs/core/net/is_macaddress_func_test.go:58]: not sure actually, I've copied this test.

File: lang/funcs/core/net/is_macaddress_func.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (purpleidea) [reply lang/funcs/core/net/is_macaddress_func.go:34]: I know ;)

File: lang/funcs/core/net/is_macaddress_func.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+		V: IsMacAddress,
+	})
+}
+
+// 00:21:cc:be:e0:6b
+
+// IsMacAddress verifies that input string is a mac address
+func IsMacAddress(input []types.Value) (types.Value, error) {
+	mac := input[0].Str()
+
```
[the remark this answers is not in the document]
Comment (purpleidea) [reply lang/funcs/core/net/is_macaddress_func.go:39]: newline

File: lang/funcs/core/net/is_macaddress_func.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (purpleidea) [reply lang/funcs/core/net/is_macaddress_func.go:54]: Here's an example:

https://play.golang.org/p/txidCnVqfai

For other maps, this can be a panic depending on what the value is. In this case it's a bool, so it's safe. But that's accidental.

File: lang/funcs/core/net/is_macaddress_func_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (purpleidea) [reply lang/funcs/core/net/is_macaddress_func_test.go:55]: I didn't understand your reply, can you rephrase?

[the remark this answers is not in the document]
Comment (S-ign) [reply lang/funcs/core/net/is_macaddress_func_test.go:55]: I see now, I initially changed everything else to return bool but this variable was named expectedStr so it didn't make sense.
I'm going to change my function so that it returns string "true" or "false" instead.

File: lang/funcs/core/net/is_macaddress_func.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (S-ign) [reply lang/funcs/core/net/is_macaddress_func.go:54]: ok I see, when assigning a map[key] to a variable go also returns a bool if it exists and you can also assign that to a variable.

File: lang/funcs/core/net/is_macaddress_func.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+)
+
+func init() {
+	simple.ModuleRegister(ModuleName, "is_macaddress", &types.FuncValue{
+		T: types.NewType("func(a str) bool"),
+		V: IsMacAddress,
+	})
+}
+
+// IsMacAddress takes a single string mac address as input, checks that it
```
Comment (purpleidea) [inline lang/funcs/core/net/is_macaddress_func.go:34]: So this comment is describing the code. And we don't really need to do that for short snippets of code, because well, the code does that. And if we ever changed one of these two, then things would be out of sync.

More importantly, what we _want_ to do, is tell the user and important things. It might be important to tell them which type of macaddresses we validate, and that sort of thing. Talking about the separators is good though.

File: lang/funcs/core/net/is_macaddress_func_test.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	for _, test := range macAddressTests {
+		t.Run(test.name, func(t *testing.T) {
+			test := test
+			output, _ := IsMacAddress([]types.Value{
+				&types.StrValue{V: test.mac},
+			})
+			expectedBool := &types.BoolValue{V: test.expected}
+
+			if err1 := output.Cmp(expectedBool); err1 != nil {
+				t.Errorf("mac: %s, expected: %s, got: %s", test.mac, expectedBool, output)
```
Comment (purpleidea) [inline lang/funcs/core/net/is_macaddress_func_test.go:59]: What is %s for? Are we using the right format string everywhere?

File: lang/funcs/core/net/is_macaddress_func_test.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+		{"invalid mac invalid delimiter", "0a;1b;2c;3d;4f;56", false, nil},
+		{"invalid mac mixed delimiters", "0a:1b:2c-3d:4f:56", false, nil},
+		{"invalid mac invalid delimiter position", "0a::1b2:c3d:4f:56", false, nil},
+		{"invalid mac invalid delimiter position", "0a--1b2-c3d-4f-56", false, nil},
+	}
+
+	for _, test := range macAddressTests {
+		t.Run(test.name, func(t *testing.T) {
+			test := test
+			output, _ := IsMacAddress([]types.Value{
```
Comment (purpleidea) [inline lang/funcs/core/net/is_macaddress_func_test.go:53]: never ignore errors without a comment saying why or a really good reason.

**Review by purpleidea — CHANGES_REQUESTED** [review]
Okay. I hope you've been learning from the review process. There's one big change we didn't make yet. I forgot if I delayed telling you on purpose, or if I forgot.

We'd definitely like to re-use existing code from the stdlib wherever possible. First, because that saves time. Second, because that code is probably tested more often. Third, because it is probably more optimized.

Unfortunately for your patch (but not for your learning experience) there's already a mac address parser in the stdlib. It may not be exactly what we want, but we probably want to use it somehow.

So have a look, find it, update this code and anything else, and please push a new version.

Use your newly git skills to combine these commits with the new code you're about to write into a single commit.

Thanks!

File: lang/funcs/core/net/is_macaddress_func_test.go (hunk 7 above)
Comment (S-ign) [reply lang/funcs/core/net/is_macaddress_func_test.go:59]: %s is for the uninterpreted bytes of the string or slice, maybe i should us %v for the string default value.

File: lang/funcs/core/net/is_macaddress_func_test.go (hunk 8 above)
Comment (S-ign) [reply lang/funcs/core/net/is_macaddress_func_test.go:53]: alright, I'll leave a comment saying that err will always return nil.
